### ISSUE TITLE
Measure asynchronous response times

### ISSUE THREAD

**Issue by willr3** [Hyperfoil/Hyperfoil#152]
We could use a way to measure the response time for asynchronous events.
For example: Hyperfoil sends an http message to service-A and measures how long it takes for the message to reach service-B.

Knative eventing is based on this asynchronous model and would like to use Hyperfoil for performance testing but I do not see how we can support that performance measurement with the current features. 

hopefully we can add this without polling or breaking the cloudevents spec.

**Comment by rvansa** [comment]
I'll expect further details regarding the implementation, ideally with a test-server (preferably Vert.x, but any container would work for me, too) that exposes the interface Hyperfoil should be using.

**Comment by willr3** [comment]
I can see this use case including starting as well as stopping a time measurement based on an asynchronous event.
I think Hyperfoil could expose a web hook that receives a cloudevent (or just http request) and uses headers or payload to identify the associated time measurment in Hyperfoil.
I think it would be cleanest to create a separate URL for each measurement (both start and stop endpoints) in the Hyperfoil script then use a message header (`Ce-id` for cloudevents) to associate the message with the Hyperfoil session.

**Comment by rvansa** [comment]
@willr3 Any news on this? Is there an example service I could run this against?

Also just saying 'cloudevents' is not enough. There are implemented over many protocols and basically just describe common headers to identify the message. I could imagine the most natural implementation in our context could be websockets.

**Comment by slinkydeveloper** [comment]
Hi @rvansa, for knative eventing use case, http (in particular binary mode) is just enough, because that's the only transport we support.

@willr3 for the "end" message, in the existing tests we used to send all the events with the same _ce-type_, except the last one with _ce-type_ `end`

**Comment by rvansa** [comment]
@slinkydeveloper Do you have an example of a setup where I could wait for the event to be published?

**Comment by slinkydeveloper** [comment]
@rvansa you can't wait for the event to be published, you can just listen for events to come (push based architecture). The performance tests I've developed for knative (they're very bad, but they should give you the idea of how it works), have a flow that looks like that (arrows are http requests):

http ce sender -> (some knative component) -> http ce receiver

(some knative component) could be:

* nothing: (aka sending to receveiver directly as baseline)
* channel: it could be kafka channel or in memory channel
* broker: broker could be mtbroker based on in memory channel and mtbroker based on kafka channel

Some sample confs:

* channel IMC: https://github.com/knative/eventing/blob/master/test/performance/benchmarks/channel-imc/100-channel-perf-setup.yaml#L88
* channel Kafka: https://github.com/knative-sandbox/eventing-kafka/blob/master/test/performance/channel-kafka/100-channel-perf-setup.yaml#L156
* broker IMC: https://github.com/knative/eventing/blob/master/test/performance/benchmarks/broker-imc/100-broker-perf-setup.yaml#L34
* broker Kafka: https://github.com/knative-sandbox/eventing-kafka/blob/master/test/performance/broker-kafka/100-broker-perf-setup.yaml#L102

At the end of the day, Hyperfoil already has a sender, it's just a matter of configuring the right http headers to "fake cloudevents". From my understanding, what's missing is a receiver, which is an http server that just receives requests, replies with 202 and parses the `ce-` headers to figure out type and id (which are the only metadata we need in the test). Using the id then, Hyperfoil should be able to correlate sent and received events.

Note: Some events might be lost in the process (in case the components in the middle are overloaded), so be sure we can timeout the test somehow and send the `end` benchmark event multiple times.

**Comment by rvansa** [comment]
We can definitely make the sender part more convenient to use, that would be a `sendCloudEvent` step, adding the `ce-` headers etc.

What was unclear is the ownership of the receiver part. As it works as a server, the CE receiver should not be part of Hyperfoil - it should be a part of the test deployment. This service accepts cloudevents, records precise timestamps (how do you expect to synchronize the time on the sender and receiver?) and publishes this information in a way that is consumable by HF connecting as client.

Hyperfoil sends the cloudevents and connects to the receiver, retrieving the timestamps and identifier and correlating those. What needs to be defined is the means of publishing the [id, timestamp] tuples.

**Comment by slinkydeveloper** [comment]
> Hyperfoil sends the cloudevents and connects to the receiver, retrieving the timestamps and identifier and correlating those. What needs to be defined is the means of publishing the [id, timestamp] tuples.

That sounds like a good approach too! I can work on the details for the receiver, a vert.x application should fit the needs

**Comment by rvansa** [comment]
I wonder if the HF component that will connect to CE receiver should be somehow aggregated. At first I've assumed that every session will query the receiver individually for the request that it just received but I am afraid that having a 1:1 ratio between cloud events and the queries would significantly affect the latency of processing the cloud events in the receiver. On the other hand we could have a component shared by all sessions in given executor that will get the updates streamed and sessions will check the timestamps for the requests only locally.

**Comment by slinkydeveloper** [comment]
@slinkydeveloper in the existing test infra we give a huge amount of memory to the receiver and we keep all the raw results in memory, which are then sent to the component that does the aggregation with the sent events. Not the most scalable solution, it doesn't work with long running benchmarks, but it does its job for small benchmarks...

Otherwise we could aggregate and each, say 10 seconds, a separate thread can send the results to hyperfoil. Otherwise, the solution to avoid this problem at all is to let hyperfoil expose the ce receiver.

**Comment by rvansa** [comment]
@slinkydeveloper That wouldn't solve the problem when you have multiple agents. If agent A sends the event and knative decides to load-balance the response to agent B we would have to move the information around. Alternatively you'd have to customize the event to always route it back to the same agent. Assuming that you want to run in a setup where knative can talk to agents at all.

Another downside (though IMO not that big) is not being able to scale the receivers independently to agents.

Keeping raw results in memory on receiver also means keeping the request start times on the sender - and that means allocation of memory (and long-lived one!) which HF tries to avoid (in order to minimize disturbances from GC).

I am also trying to avoid receiver connecting to HF at all. It's HF that should connect to receiver (since that it publicly addressable). Then receiver would stream data to HF (HF acking what can be forgotten). Protocol is up to you - plain TCP is fine IMO, websockets would do as well.

**Comment by slinkydeveloper** [comment]
> That wouldn't solve the problem when you have multiple agents. If agent A sends the event and knative decides to load-balance the response to agent B we would have to move the information around. Alternatively you'd have to customize the event to always route it back to the same agent. Assuming that you want to run in a setup where knative can talk to agents at all.

TBH I'm a bit lost here, wdym by agent? services?  If that's the case, you can tell where the event should be routed (we do that in the existing performance tests).
Or are you referring to the system which records and aggregates values?

> Keeping raw results in memory on receiver also means keeping the request start times on the sender - and that means allocation of memory (and long-lived one!) which HF tries to avoid (in order to minimize disturbances from GC).

In your mind, are you thinking to a test with more than one receiver? Can you try to explain it with a quick graph? :smile: 

> Protocol is up to you - plain TCP is fine IMO, websockets would do as well.

Websockets are not free, raw tcp is better i think

**Comment by rvansa** [comment]
> > That wouldn't solve the problem when you have multiple agents. If agent A sends the event and knative decides to load-balance the response to agent B we would have to move the information around. Alternatively you'd have to customize the event to always route it back to the same agent. Assuming that you want to run in a setup where knative can talk to agents at all.
> 
> TBH I'm a bit lost here, wdym by agent? services? If that's the case, you can tell where the event should be routed (we do that in the existing performance tests).
> Or are you referring to the system which records and aggregates values?

Hyperfoil is a distributed load driver. *Agents* are the processes that send (HTTP) requests to the tested system (execute scenario) and send statistics to controller. *Controller* starts the agents, orchestrates the test, gathers statistics etc. See https://hyperfoil.io/docs/concepts.html

> 
> > Keeping raw results in memory on receiver also means keeping the request start times on the sender - and that means allocation of memory (and long-lived one!) which HF tries to avoid (in order to minimize disturbances from GC).
> 
> In your mind, are you thinking to a test with more than one receiver? Can you try to explain it with a quick graph? smile

When you invoke a godzillion of events per second one instance of a service certainly won't be able to handle them, this will have to scale out. In k8s you'll probably have these load-balanced through a Service (however we'll probably need another headless service to enumerate them all and connect each agent to each receiver).

> 
> > Protocol is up to you - plain TCP is fine IMO, websockets would do as well.
> 
> Websockets are not free, raw tcp is better i think

AFAIK Websockets only give you framing on top of TCP, so besides the initial connection upgrade the difference should be negligible (I don't have any data to back that claim, through). I would chose based on convenience and number of wheels reinvented :)

**Comment by slinkydeveloper** [comment]
> Hyperfoil is a distributed load driver. Agents are the processes that send (HTTP) requests to the tested system (execute scenario) and send statistics to controller. Controller starts the agents, orchestrates the test, gathers statistics etc. See https://hyperfoil.io/docs/concepts.html

Went through this page, now I see your POV. So I guess, following the same concepts, receiver should be themselves agents created and managed by hyperfoil, sending raw data to hyperfoil controller to process them. Does that makes sense? Are you thinking to something like this?

![Untitled Diagram(1)](https://user-images.githubusercontent.com/6706544/110314213-dca72500-8007-11eb-91f1-cabf47470f13.png)

As you sad, the load is "balanced" using a k8s svc (although at some point we'll need knative svcs, but let's keep them out of the picture for now), so each receiver will receive separate messages, which is ok.

Each agent (both sender and receiver) should be connected to the hyperfoil controller, send them the raw data and let the controller aggregate them. Or we have a separate system where we send all the raw data, we let it aggregate the data and then send the results back to hyperfoil (although, I don't see the point of doing this, why not just let the hyperfoil controller do this?)

I'm sorry if I'm stating the obvious, I'm trying to picture myself the overall architecture of this test.

> When you invoke a godzillion of events per second one instance of a service certainly won't be able to handle them, this will have to scale out. In k8s you'll probably have these load-balanced through a Service (however we'll probably need another headless service to enumerate them all and connect each agent to each receiver).

I'm pretty sure we won't reach these numbers... Let's start by reasoning on a benchmark that pushes 1000 rps :smile:

**Comment by rvansa** [comment]
Please do not overload the term *Agent* for a process that does not execute the Scenario. That would only confuse things.
The controller does not consume raw data for each request, only aggregated results (think a histogram for the last second, for each metric =~= endpoint being hit). Controller aggregates these aggregates from multiple agents. Single process cannot have the bandwidth to handle individual requests, we are trying to avoid bottlenecks.

I would prefer to draw a line between this particular test and Hyperfoil as a generic tool. In order to not overcomplicate things Hyperfoil has an explicit non-goal to not be test deployer, so I'd like to refrain from adding a generic tooling to start some generic services - the receiver in your case. In my view this is a part of a higher tier in the 'test' that sets up knative, instructs Hyperfoil to run the test etc.

> I'm pretty sure we won't reach these numbers... Let's start by reasoning on a benchmark that pushes 1000 rps

So you're saying Knative is not web-scale? :) Jokes aside, let's not waste time with implementing an architecture that you'd have to rework anyway after the initial POC. When I think about this more I believe that the receivers will eventually need to know which agent did each event originate from because we don't want to inform all agents about every request.

**Comment by slinkydeveloper** [comment]
> In my view this is a part of a higher tier in the 'test' that sets up knative, instructs Hyperfoil to run the test etc.

That's ok for me, I don't mind where which component lives, we can think about it later. I'm just trying to grasp the overall architecture of the test :)

> Controller aggregates these aggregates from multiple agents. Single process cannot have the bandwidth to handle individual requests, we are trying to avoid bottlenecks.

I see, so, from your POV, who does the actual aggregation? The agent who sent the event?

> When I think about this more I believe that the receivers will eventually need to know which agent did each event originate from because we don't want to inform all agents about every request.

I believe we can easily add some agent id/host just adding it to the sent cloudevent. The receiver, looking at that particular cloudevent attribute

**Comment by rvansa** [comment]
> > In my view this is a part of a higher tier in the 'test' that sets up knative, instructs Hyperfoil to run the test etc.
> 
> That's ok for me, I don't mind where which component lives, we can think about it later. I'm just trying to grasp the overall architecture of the test :)
> 
> > Controller aggregates these aggregates from multiple agents. Single process cannot have the bandwidth to handle individual requests, we are trying to avoid bottlenecks.
> 
> I see, so, from your POV, who does the actual aggregation? The agent who sent the event?

Yes. Agent can diff send time and receive time and record that into statistics (which are later sent to the controller).

Hmm, I realized one thing, though. Since we're already sending custom data (at least some unique request ID and agent+thread ID) and the receiver is custom implemented there's really a little point in matching the request and reception in agents: we could replace request ID by the very timestamp and do it all in the receiver. My apologies for contradicting myself - I've been trying to graft this test onto existing architecture, but a rethink might be simpler in the end.

So despite I've refused the idea before lunch it might actually simplify things a lot if we could let receivers join the Vert.x eventbus that Hyperfoil uses, and allow them to send statistics to the controller directly. The controller would not start the receivers at all and it would not track them either - just accept any metric these would produce and store it along.

Obviously this requires the receivers to reside in the same network as Hyperfoil - I guess that's not limiting. I wonder if there's any further development of this benchmark you could foresee that would rely on matching requests, or e.g. invoking another event only after the first event was processed.

**Comment by rvansa** [comment]
Without the matching we would have to manually verify that the number of requests sent and received does match. I can see a problem with events that are received after the *phase* completed on agents - controller would not wait for the receivers and we'd probably miss a good deal of the events that were sent at the end. Unless we track the receivers.

**Comment by slinkydeveloper** [comment]
> So despite I've refused the idea before lunch it might actually simplify things a lot if we could let receivers join the Vert.x eventbus that Hyperfoil uses, and allow them to send statistics to the controller directly

If the receiver is vert.x, then it's easy peazy to do that, no need for custom protocols etc. So in your idea, the receiver will perform the same job of the agent: compute raw values, aggregate them and then publish to the controller through the event bus, right?
In order to do that, can we build the receiver starting depending directly on the code of the agent? This way, we'll just need to implement the http receiver and the aggregation.

>  I wonder if there's any further development of this benchmark you could foresee that would rely on matching requests, or e.g. invoking another event only after the first event was processed.

I don't think we need that, we're interested in recording only e2e latencies.

>  I can see a problem with events that are received after the phase completed on agents - controller would not wait for the receivers and we'd probably miss a good deal of the events that were sent at the end

We solve this problem in the existing benchmark by not solving it: a timeout to wait for the aggregated events from the receiver does the trick.

**Comment by rvansa** [comment]
> In order to do that, can we build the receiver starting depending directly on the code of the agent? This way, we'll just need to implement the http receiver and the aggregation.

Sure, Hyperfoil is published in Maven repository: https://mvnrepository.com/artifact/io.hyperfoil/hyperfoil-clustering

> We solve this problem in the existing benchmark by not solving it: a timeout to wait for the aggregated events from the receiver does the trick. If the events were not received for that period of time, the event is considered lost

Yes, timeout it certainly easier that tracking, though it will have to be an another configuration option that's a bit unclear in the normal benchmark. Anyway it's probably the least intrusive way.

**Comment by slinkydeveloper** [comment]
Cool so, how do we wanna proceed? Do you have any ideas in mind on how to go forward? Do we want to meet to come out with an implementation plan?

**Comment by rvansa** [comment]
I think that the statistics integration part would be up to me. It would be nice if you could
1) create a benchmark that will send the CEs;
2) create the receiver as Vert.x app that will record the latencies into a HdrHistogram (no need to depend on HF at this point) and print the number of events and average latency to stdout every second.
3) test that this still works fine if you add knative to the sauce

I can take that over then and make it send the stats to the controller. Does that sound viable?

**Comment by slinkydeveloper** [comment]
@rvansa I'm working on my part of the issue, the WIP is here: https://github.com/slinkydeveloper/eventing-hyperfoil-benchmark

I'm struggling with adding the timestamp in milliseconds in an header (https://github.com/slinkydeveloper/eventing-hyperfoil-benchmark/blob/main/ce-requests.hf.yaml#L14), I was looking in the doc but I can't find any step that does it.

Is there a built-in way to do that in hyperfoil? Or do I need to create an extension for it?

**Comment by rvansa** [comment]
Regrettably there's no step that could generate that ATM. If you create one I would put it straight to core - it seems to me that this is a generic-enough functionality. There's no technical difference between 'built-in' and 'extension' - it's just a piece of code (plus service record in META-INF/services) that is on the classpath.
This should get you started: https://hyperfoil.io/quickstart/quickstart8.html

**Comment by slinkydeveloper** [comment]
> There's no technical difference between 'built-in' and 'extension' - it's just a piece of code (plus service record in META-INF/services) that is on the classpath.

Sounds good, I'll do that in my fork and if it works I'll PR it to hyperfoil :)

**Comment by slinkydeveloper** [comment]
@rvansa I think I'm good to go and now it's your time to figure out the rest :)

* I've created a PR to add the timestamp step https://github.com/Hyperfoil/Hyperfoil/pull/173
* The receiver works, I've added some yamls to configure a test environment with InMemoryChannel https://github.com/slinkydeveloper/eventing-hyperfoil-benchmark/tree/main/imc-test-config and I did a local test letting messages go through IMC and it works.
* The hf benchmark definition is definitely drafted. I left some questions in the TODOs where I need your help.

Let me know if there's something else I need to help with

**Comment by rvansa** [comment]
Hi @slinkydeveloper I've pushed my changes in https://github.com/rvansa/eventing-hyperfoil-benchmark
This obviously depends on snapshot version of Hyperfoil (I could release later this week). Note the `000-hyperfoil-cluster.yaml` service definition - I'll add automatic creation of this into hyperfoil operator.

Some of the last events are not reported - Hyperfoil receives the stats but decides to ignore them because the phase is already marked as complete and the statistics have been analyzed. I have considered querying the receiver for completion but as it doesn't know when it receives all events it would have to just block for an arbitrary amount of time. Extending the phase this way would skew any place that calculates throughput, so I've decided rather to let it miss a fraction of results.
